numpy/core/src/npymath/npy_math_private.h

```diff
@@ -172,8 +172,10 @@ do {                                                            \
   (d) = sf_u.value;                                             \
 } while (0)
 
-#ifdef NPY_USE_C99_COMPLEX
-#include <complex.h>
+#if defined(__cplusplus) && defined(__FCC_version__) && defined(HAVE_SSL2)
+  #include <complex> 
+#elif defined(NPY_USE_C99_COMPLEX)
+  #include <complex.h>
```

Comment:
This again looks like you are using a C++ compiler, that seems not ideal?

Reply:
As a test, in the case of C++, I changed it to not include complex and built it with the gcc compiler, which resulted in an error.
```
#if defined(NPY_USE_C99_COMPLEX) && !defined(__cplusplus)
  #include <complex.h>
#endif
```
This is part of the build.log.
```
INFO: compiling C++ sources
INFO: C compiler: /opt/ml/tools/gcc-12.1.0_gcc-11.3.0/bin/g++ -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC
creating build/temp.linux-aarch64-3.10/numpy/core/src/npysort
INFO: compile options: '-DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DHAVE_CBLAS -Ibuild/src.linux-aarch64-3.10/numpy/core/src/multiarray -Ibuild/src.linux-aarch64-3.10/numpy/core/src/common -Ibuild/src.linux-aarch64-3.10/numpy/core/src/umath -Inumpy/core/include -Ibuild/src.linux-aarch64-3.10/numpy/core/include/numpy -Ibuild/src.linux-aarch64-3.10/numpy/distutils/include -Ibuild/src.linux-aarch64-3.10/numpy/core/src/npysort -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/src/_simd -I/home/yamadaf2/numpy-sve/.venv-3.10-221004/include -I/opt/ml/tools/Python-3.10.7_gcc-11.3.0/include/python3.10 -Ibuild/src.linux-aarch64-3.10/numpy/core/src/common -Ibuild/src.linux-aarch64-3.10/numpy/core/src/npymath -c'
extra options: '-std=c++11 -D__STDC_VERSION__=0 -fno-exceptions -fno-rtti'
INFO: g++: numpy/core/src/npysort/quicksort.cpp
INFO: g++: numpy/core/src/npysort/mergesort.cpp
INFO: g++: numpy/core/src/npysort/timsort.cpp
INFO: g++: numpy/core/src/npysort/heapsort.cpp
INFO: g++: numpy/core/src/npysort/radixsort.cpp
INFO: g++: numpy/core/src/npysort/binsearch.cpp
INFO: g++: numpy/core/src/npysort/selection.cpp
INFO: g++: numpy/core/src/multiarray/textreading/tokenize.cpp
In file included from numpy/core/src/npymath/npy_math_internal.h.src:56,
                 from numpy/core/include/numpy/npy_math.h:568,
                 from numpy/core/src/npysort/selection.cpp:18:
numpy/core/src/npymath/npy_math_private.h:520:14: error: ‘complex’ in namespace ‘std’ does not name a template type
  520 |           std::complex<double> c99_z;
      |                ^~~~~~~
numpy/core/src/npymath/npy_math_private.h:35:1: note: ‘std::complex’ is defined in header ‘<complex>’; did you forget to ‘#include <complex>’?
   34 | #include "numpy/npy_endian.h"
  +++ |+#include <complex>
   35 | #include "numpy/npy_common.h"
numpy/core/src/npymath/npy_math_private.h:529:14: error: ‘complex’ in namespace ‘std’ does not name a template type
  529 |            std::complex<float> c99_z;
      |                 ^~~~~~~
numpy/core/src/npymath/npy_math_private.h:529:9: note: ‘std::complex’ is defined in header ‘<complex>’; did you forget to ‘#include <complex>’?
  529 |            std::complex<float> c99_z;
      |            ^~~
numpy/core/src/npymath/npy_math_private.h:538:14: error: ‘complex’ in namespace ‘std’ does not name a template type
numpy/core/src/npymath/npy_math_private.h:538:9: note: ‘std::complex’ is defined in header ‘<complex>’; did you forget to ‘#include <complex>’?
  538 |         std::complex<long double> c99_z;
      |         ^~~
INFO: g++: numpy/core/src/umath/clip.cpp
In file included from numpy/core/src/npymath/npy_math_internal.h.src:56,
                 from numpy/core/include/numpy/npy_math.h:568,
                 from numpy/core/include/numpy/halffloat.h:5,
                 from numpy/core/src/umath/clip.cpp:11:
```

Reply:
The file should be used both in C++ and C99 code, I am confused why it would fail for you only in C++ mode, e.g. MSVC has no `complex.h` (thus that guard), but right now we never use `<complex>` (which MSVC does have)?

Maybe it makes sense to use `<complex>` when in C++, but it seems strange to do it compiler specific and is confusing why it should be needed for fujitsu when it isn't for the others.

Reply:
Thank you for comments.
Sorry, in `site.cfg`, I had specified the wrong paths for `library_dirs` and `include_dirs` for SSL2, which caused the compile error. I confirmed that if I change these to the correct paths, I can build with the original code.